### ISSUE TITLE
Changes to Gateway objects affect other Gateway objects in different namespaces

### ISSUE THREAD

**Issue by HXK123** [istio/istio#12336]
**Describe the bug**
Changes to `Gateway` objects affect other `Gateway` objects in different namespaces.

When two `Gateway` objects are defined in separate namespaces with `spec.selector.istio: ingressgateway` but both having different `port` dictionaries defined, changes to one `Gateway` object seem to also affect the other.

Example:

_Service 001_ on Kubernetes namespace `ns001`:

```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: mygateway
spec:
  selector:
    istio: ingressgateway # use istio default ingress gateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
  - port:
      number: 443
      name: https
      protocol: HTTPS
    tls:
      mode: SIMPLE
      serverCertificate: /etc/istio/ingressgateway-certs/tls.crt
      privateKey: /etc/istio/ingressgateway-certs/tls.key
    hosts:
    - "*"
---
kind: VirtualService
metadata:
  name: httpbin
spec:
  hosts:
  - httpbin.example.com
  gateways:
  - mygateway
  http:
  - match:
    - uri:
        prefix: /status
    - uri:
        prefix: /delay
    route:
    - destination:
        port:
          number: 8000
        host: httpbin
```

_Service 002_ on namespace `ns002`:

```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: mygateway
spec:
  selector:
    istio: ingressgateway # use istio default ingress gateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - "*"
---
kind: VirtualService
metadata:
  name: bookinfo
spec:
  hosts:
  - bookinfo.example.com
  gateways:
  - mygateway
  http:
  - match:
    - uri:
        prefix: /status
    - uri:
        prefix: /delay
    route:
    - destination:
        port:
          number: 8000
        host: bookinfo
```

After deploying both sets of `Gateway` and `VirtualService` objects, https://httpbin.example.com/status works as expected, but now https://bookinfo.example.com/status started working as well.

**Expected behavior**
`Gateway` objects with differing `port` dictionaries should be respected within their own namespaces.  In the example above, https://bookinfo.example.com/status should fail as only HTTP traffic is allowed.

**Steps to reproduce the bug**
1. Create _Service 001_ in the indicated namespace from the above examples.
2. Verify that https://httpbin.example.com/status works.
3. Create _Service 002_ in the indicated namespace from the above examples.
4. Test the endpoint at https://bookinfo.example.com/status and review the results.

**Version**
```
istioctl version
Version: 1.0.5
GitRevision: c1707e45e71c75d74bf3a5dec8c7086f32f32fad
User: root@6f6ea1061f2b
Hub: docker.io/istio
GolangVersion: go1.10.4
BuildStatus: Clean
```

**Installation**
* Via `helm install` on `istio-1.0.5/install/kubernetes/helm`.

**Environment**
* Google Cloud Platform's Google Kubernetes Engine.

**Cluster state**
* N/A.

**Comment by edussx** [comment]
This looks like a similar issue that I ran into: https://github.com/istio/istio/issues/10313

**Comment by HXK123** [comment]
Thanks for responding @edussx.  Is the recommendation to use unique `port` names to increase isolation?

Would this "merge" behaviour also explain why removing settings from one `port` does not seem to have an effect on other `port`s?  For example, consider these two `Gateway`s created on different namespaces:

Namespace _ns001_:

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: mygateway
spec:
  selector:
    istio: ingressgateway # use istio default ingress gateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    tls:
      httpsRedirect: true # sends 301 redirect for http requests
    hosts:
    - "*"
```

Namespace _ns002_:

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: mygateway
spec:
  selector:
    istio: ingressgateway # use istio default ingress gateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - "*"
```

Originally, we had tls.httpRedirect for both `Gateway`s, but chose to remove it for the _ns002_ version of the `Gateway` object.  However, the `VirtualService` tied to the `Gateway` on _ns002_ still redirected traffic from HTTP to HTTPS even though the configuration was missing.  Is that expected with this configuration "merge" behaviour given the "http" `port`s have the same names?

Thanks!

**Comment by rshriram** [comment]
fixed by #12663
